### PR TITLE
feat: math function vectorization

### LINKED ISSUE #5352: Math Function Vectorization
#### Current Code Base
TiFlash currently has an relatively old math function vectorization mechanism inheritated from ClickHouse's old code base:
https://github.com/pingcap/tiflash/blob/34d7570a34c0daec5ca9c705296470b31d743d91/dbms/src/Functions/FunctionsMath.h#L30

However, when distributing the products, we never tried to enable it. Actually, at ClickHouse's master, this library has already been removed:
https://github.com/ClickHouse/ClickHouse/blob/d0e6f8ebc120a682bb09231fbbba1a904c8c29bb/src/Functions/FunctionMathUnary.h#L171

It seems that [`vectorclass`](https://github.com/vectorclass/version2) has already become somehow inactive since 2021. Clickhouse quit using vectorized math functions anymore.

#### Vectorization Solution

Besides `vectorclass`, there are several alternatives providing math function vectorization:

- [`SLEEF`](https://sleef.org/) is a promising candidate. However, there is no new release since 2020.
- [`SVML`](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/intrinsics/intrinsics-for-short-vector-math-library-ops.html) is the most famous one. However, it is designed for Intel CPU and it is a proprietary product.
- [`LIBMVEC`](https://sourceware.org/glibc/wiki/libmvec) is a part of glibc licenced under LGPL. TiFlash is now mostly used on GNU/Linux envrionment. This can be considered as a suitable choice.

For aarch64 platform, tiflash has already introduced [optimized-routines](https://github.com/ARM-software/optimized-routines) as one of its contrib library; which can provide vectorized variants for `log`, `cos`, `sin`, `exp`.

### Proposed Solution
`libmvec` is generally available in environment with `glibc 2.22+`. However, as we still need to support some ancient platform like `CentOS 7`, it is possible that the deployed env does not have `libmvec`. 

Therefore, we can use `dlopen` with `dlsym` to dynamically detect whether `libmvec` is available.

Another consideration is that vectorized math function will have a lower precision. Therefore, there are potential changes in behavior when vectorization is enabled. To avoid trouble, we should provide a switch for the user and disable vectorization on default.

### Summary

- remove `dbms/cmake/find_vectorclass.cmake` and related macros in source.
- add `unary_math_vectorization = false` to tiflash's configuration.
- add unit tests for unary math functions.
- vectorized the following functions on `x86-64` using `libmvec`(`libmvec`'s `exp` performs slower, so we do not use it):
   - `cos`
   - `sin`
   - `tan`
   - `acos`
   - `asin`
   - `atan`
   - `log`
   - `log2`
   - `cbrt`
- vectorize the following functions on `aarch64` using `optimized-routines`:
   - `cos`
   - `sin`
   - `log`
   - `exp`

#### How to check `libmvec` is available?
image
#### Vector ABI
image
[more details](https://sourceware.org/glibc/wiki/libmvec?action=AttachFile&do=view&target=VectorABI.txt)

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
Signed-off-by: Schrodinger ZHU Yifan <i@zhuyi.fan>

### What problem does this PR solve?

Issue Number: close #5352 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/UnaryMath/AMD64.h [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    static constexpr auto & load = Load;
+    static constexpr auto & store = Store;
+
+    UnaryFuncPtr funcs[static_cast<size_t>(LibMVecFunc::sentinel)]{};
+
+    MVec(const char * prefix, void * handle)
+    {
+        char sym_buffer[64]{};
+        auto length = strlen(prefix);
+        std::memcpy(sym_buffer, prefix, length);
```
[the remark this answers is not in the document]
Comment (SchrodingerZhu) [reply dbms/src/Functions/UnaryMath/AMD64.h:99]: It should not be possible here. But I had better leave a remark.

File: dbms/src/Functions/FunctionsMath.h [hunk 2]
```diff
@@ -716,38 +666,135 @@ struct PowName
...
+            dst[0] = static_cast<Float64>(::FUNC(static_cast<Float64>(src[0])));    \
+        }                                                                           \
+    }
+
+#define UNARY_FUNCTION_IMPL_NORMAL(X, FUNC) \
+    using X##FunctionImpl = UnaryFunctionPlain<X##Name, FUNC>
+
+#if defined(__linux__) && defined(__aarch64__)
+UNARY_FUNCTION_IMPL(Log, log);
+UNARY_FUNCTION_IMPL_NORMAL(Log2, log2);
```
Comment (JaySon-Huang) [inline dbms/src/Functions/FunctionsMath.h:693]: add comments to explain why we only make parts of the functions be vectorized? This paragraph is copied from the issue https://github.com/pingcap/tiflash/issues/5352
```
- vectorized the following functions on `x86-64` using `libmvec`(`libmvec`'s `exp` performs slower, so we do not use it):
   - `cos`
   - `sin`
   - `tan`
   - `acos`
   - `asin`
   - `atan`
   - `log`
   - `log2`
   - `cbrt`
- vectorize the following functions on `aarch64` using `optimized-routines`:
   - `cos`
   - `sin`
   - `log`
   - `exp`
```

File: dbms/src/Functions/FunctionsMath.h [hunk 3]
```diff
@@ -716,38 +666,102 @@ struct PowName
...
+    {                                                                            \
+        static constexpr auto name = X##Name::name;                              \
+        static constexpr auto rows_per_iteration = 1;                            \
+        template <typename T>                                                    \
+        static void transform(Float64 * dst, const T * src, size_t size)         \
+        {                                                                        \
+            ::DB::UnaryMath::FUNC##Transform<T>(dst, src, size);                 \
+        }                                                                        \
+        template <typename T>                                                    \
+        static void execute(const T * src, Float64 * dst)                        \
```
[the remark this answers is not in the document]
Comment (SchrodingerZhu) [reply dbms/src/Functions/FunctionsMath.h]: fixed

File: dbms/src/Functions/FunctionsMath.h [hunk 4]
```diff
@@ -716,38 +666,102 @@ struct PowName
...
+using FunctionTan = FunctionMathUnaryFloat64<TanFunctionImpl>;
+using FunctionAsin = FunctionMathUnaryFloat64<AsinFunctionImpl>;
+using FunctionAcos = FunctionMathUnaryFloat64<AcosFunctionImpl>;
+using FunctionAtan = FunctionMathUnaryFloat64<AtanFunctionImpl>;
+using FunctionErf = FunctionMathUnaryFloat64<ErfFunctionImpl>;
+using FunctionErfc = FunctionMathUnaryFloat64<ErfcFunctionImpl>;
 using FunctionLGamma = FunctionMathUnaryFloat64<UnaryFunctionPlain<LGammaName, std::lgamma>>;
 using FunctionTGamma = FunctionMathUnaryFloat64<UnaryFunctionPlain<TGammaName, std::tgamma>>;
-using FunctionPow = FunctionMathBinaryFloat64<BinaryFunctionVectorized<PowName, pow>>;
+using FunctionPow = FunctionMathBinaryFloat64<BinaryFunctionPlain<PowName, pow>>;
```
[the remark this answers is not in the document]
Comment (SchrodingerZhu) [reply dbms/src/Functions/FunctionsMath.h]: fixed

File: dbms/src/Functions/UnaryMath/AMD64.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (SchrodingerZhu) [reply dbms/src/Functions/UnaryMath/AMD64.h]: fixed

File: dbms/src/Functions/FunctionsMath.h (hunk 2 above)
Comment (SchrodingerZhu) [reply dbms/src/Functions/FunctionsMath.h:693]: added comments

**Comment by SchrodingerZhu** [comment]
/hold

there is some extra changes to be examined on aarch64

**Pull request closed without merging**